Add image optimization for imported attachments

pfefferle opened the pull request:
## Proposed changes:

* Add automatic image optimization for imported ActivityPub attachments to reduce file sizes
* Resize large images to max 1200px dimension while maintaining aspect ratio
* Convert images to WebP format when supported by the server (falls back to JPEG when not)
* Skip GIFs (may be animated) and SVGs (vector graphics) to preserve their special properties

### Other information:

- [x] Have you written new tests for your changes, if applicable?

Existing tests were updated to handle the new WebP conversion behavior.

## Testing instructions:

* Set up a local WordPress environment with wp-env
* Have another ActivityPub-enabled instance (like Mastodon) post an image larger than 1200px
* Follow that account and receive the post
* Check the imported image in `wp-content/uploads/activitypub/` - it should be:
  * Resized to max 1200px on the longest edge
  * Converted to WebP format (or JPEG if WebP not supported)
  * Significantly smaller file size than the original

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Add image optimization for imported attachments (resize to 1200px max, convert to WebP).

</details>

pfefferle committed:
Add image optimization for imported attachments

Optimize imported ActivityPub images by resizing large images and
converting to WebP format for reduced file sizes:

- Add MAX_IMAGE_DIMENSION constant (1200px)
- Add optimize_image() method using WordPress image editor
- Convert images to WebP when supported, fallback to JPEG
- Skip GIFs (may be animated) and SVGs (vector graphics)
- Update save_file() and save_attachment() to optimize images
- Update tests to handle WebP conversion

matticbot committed:
Add changelog

pfefferle committed:
Update changelog for image optimization feature

Change significance level from patch to minor and add image optimization details.

pfefferle commented:
@obenland in addition to #2608, I think it makes sense to also reduce the size an image format for remote images. I added the optimization to both, the imports and the remote posts. I am happy to have it for remote posts only though, because you might want to have the original image when importing your Mastodon posts.

pfefferle committed:
Fix optimize_image logic for edge cases

Address Copilot feedback:
- Early return when no resize needed for PNG/WebP (avoids returning true)
- Simplify result handling - editor->save() always returns array
- Set changed=true when any save operation occurs

pfefferle committed:
Add max_dimension parameter to optimize_image method

Allow callers to specify the maximum dimension for image resizing.

pfefferle committed:
Add max_dimension parameter to save_file method

Allow callers to specify custom image dimensions when saving files,
enabling reuse for different contexts like avatars.

obenland commented on includes/class-attachments.php:
```diff
@@ -551,6 +581,96 @@ private static function save_file( $attachment_data, $object_id, $object_type )
+		if ( ! $mime_type || ! \str_starts_with( $mime_type, 'image/' ) ) {
+			return array(
+				'path'    => $file_path,
+				'changed' => false,
```
Let's remove the "changed" information and only return the file path. It shouldn't matter whether it changed or not.

pfefferle replied:
this was an initial feedback, because it will always be handled as changed otherwise!?

obenland replied:
What are the consequences of assuming they changed when they didn't?

obenland commented on includes/class-attachments.php:
```diff
@@ -444,15 +451,31 @@ private static function save_attachment( $attachment_data, $post_id, $author_id
+		}
+
 		$file_array = array(
-			'name'     => \basename( \wp_parse_url( $attachment_data['url'], PHP_URL_PATH ) ),
```
Can we use `$tmp_file` here instead of `$attachment_data['url']`? That would allow us to remove the added 7 lines above.

pfefferle replied:
We still need to derive the filename from the original URL because `wp_tempnam()` creates files with random suffixes (e.g., `test.jpg12345`) that don't have proper extensions. I've simplified the code by:

1. Moving the filename derivation earlier
2. Renaming the temp file to have a proper extension before optimization
3. Then updating only the extension after optimization

This ensures `wp_check_filetype()` can properly detect the mime type while keeping the original filename for WordPress.

obenland commented on includes/class-attachments.php:
```diff
@@ -444,15 +451,31 @@ private static function save_attachment( $attachment_data, $post_id, $author_id
+		$mime_type = $optimized['changed'] ? '' : ( $attachment_data['mediaType'] ?? '' );
 		$post_data = array(
-			'post_mime_type' => $attachment_data['mediaType'] ?? '',
+			'post_mime_type' => $mime_type,
```
Let's just remove this and let WP handle it? I've not checked how it works under the hood, this is just based on the comment in line 475.

pfefferle committed:
Simplify optimize_image and address PR feedback

- Return only file path from optimize_image instead of array with changed flag
- Let WordPress auto-detect mime type instead of manually setting it
- Add temp file extension handling for proper mime type detection
- Add get_unique_path helper to prevent file collisions during format conversion
- Move WP_Filesystem initialization to function start

pfefferle committed:
Add tests for optimize_image and get_unique_path functions

- Test skipping non-image files (audio, video)
- Test skipping GIF files (may be animated)
- Test WebP conversion when supported
- Test resizing large images
- Test preserving small image dimensions
- Test handling non-existent files gracefully
- Test PNG file handling
- Test get_unique_path generates unique filenames

obenland reviewed (approved):
Let's get this in for now.

If you don't mind, I'd like to take a stab at iterating on this, it feels like there's some room for simplification across the class.

Unique filenames is one of them. I was about to suggest using `wp_unique_filename()` instead of `get_unique_path()` and then realized that `save_file()` also has it own way of creating a unique name. Stuff like that.
